pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -572,6 +572,46 @@ void MultiTopicsConsumerImpl::acknowledgeAsync(const MessageId& msgId, ResultCal
     }
 }
 
+void MultiTopicsConsumerImpl::acknowledgeAsync(const MessageIdList& messageIdList, ResultCallback callback) {
+    if (state_ != Ready) {
+        callback(ResultAlreadyClosed);
+        return;
+    }
+
+    std::map<std::string, MessageIdList> topicToMessageId;
+    for (const MessageId& messageId : messageIdList) {
+        auto topicName = messageId.getTopicName();
+        if (topicToMessageId.count(topicName) == 0) {
+            topicToMessageId.emplace(topicName, std::vector<MessageId>());
+        }
+        topicToMessageId[topicName].emplace_back(messageId);
+    }
+
+    std::shared_ptr<std::atomic<int>> needCallBack =
+        std::make_shared<std::atomic<int>>(topicToMessageId.size());
+    auto cb = [callback, needCallBack](Result result) {
+        if (result != ResultOk) {
+            callback(result);
+        }
+        needCallBack->fetch_sub(1);
+        if (needCallBack->load() == 0) {
+            callback(result);
+        }
+    };
+    for (const auto& kv : topicToMessageId) {
+        auto optConsumer = consumers_.find(kv.first);
+        if (optConsumer.is_present()) {
+            for (const auto& msgId : kv.second) {
+                unAckedMessageTrackerPtr_->remove(msgId);
+            }
```

Comment:
The same as my comment before, calling `remove` in a loop is not thread safe for `unAckedMessageTrackerPtr_`.